plugins/VRPNPlugin/VRPNPlugin.cpp

```diff
@@ -0,0 +1,137 @@
+
+constexpr std::array<double, 3> to_array_3d(const vrpn_float64* pos)
+{
+    return {pos[0], pos[1], pos[2]};
+}
+inline std::array<double, 3> to_array_3d(const Vector3d& v)
+{
+    return {v.x(), v.y(), v.z()};
+}
+
+Vector3d rotateVectorByQuat(const Vector3d& v, const vrpn_float64* q)
+{
+    const auto u = Vector3d{q[0], q[1], q[2]}; // vector part of the quaternion
+    const auto s = q[3];                       // scalar part of the quaternion
+
+    return 2.0 * vmml::dot(u, v) * u + (s * s - vmml::dot(u, u)) * v +
+           2.0 * s * vmml::cross(u, v);
+}
+
+PropertyMap::Property getOpenDeckPositionProperty()
+{
+    PropertyMap::Property openDeckPosition{openDeckPositionProp,
+                                           openDeckPositionProp,
+                                           openDeckInitPos};
+    openDeckPosition.markReadOnly();
+    return openDeckPosition;
+}
+
+PropertyMap::Property getOpenDeckCameraDUProperty()
+{
+    PropertyMap::Property openDeckCamDU{openDeckCamDUProp, openDeckCamDUProp,
+                                        to_array_3d(openDeckRightDirection)};
+    openDeckCamDU.markReadOnly();
+    return openDeckCamDU;
+}
+
+PropertyMap::Property getStereoModeProperty()
+{
+    return {"stereoMode",
+            "Stereo mode",
+            (int)ospray::PerspectiveCamera::StereoMode::OSP_STEREO_SIDE_BY_SIDE,
+            {"None", "Left eye", "Right eye", "Side by side"}};
+}
+
+PropertyMap::Property getInterpupillaryDistanceProperty()
+{
+    return {"interpupillaryDistance", "Eye separation", 0.0635, {0.0, 10.0}};
+}
+
+PropertyMap getDefaultCameraProperties()
+{
+    PropertyMap properties;
+    properties.setProperty(getOpenDeckPositionProperty());
+    properties.setProperty(getOpenDeckCameraDUProperty());
+    properties.setProperty(getStereoModeProperty());
+    properties.setProperty(getInterpupillaryDistanceProperty());
+    return properties;
+}
+
+void trackerCallback(void* userData, const vrpn_TRACKERCB t)
+{
+    auto camera = static_cast<Camera*>(userData);
+
+    const auto cameraDU = rotateVectorByQuat(openDeckRightDirection, t.quat);
+
+    camera->updateProperty(openDeckPositionProp, to_array_3d(t.pos));
+    camera->updateProperty(openDeckCamDUProp, to_array_3d(cameraDU));
+}
+}
+
+VRPNPlugin::VRPNPlugin(PluginAPI* api, const std::string& vrpnName)
+    : _camera{api->getCamera()}
+    , _vrpnTracker{vrpnName.c_str()}
+{
+    if (_vrpnTracker.register_change_handler(&_camera, trackerCallback,
+                                             trackedSensorId) == -1)
+    {
+        throw std::runtime_error("VRPN couldn't connect to: " + vrpnName);
+    }
+    _camera.setProperties(cameraType, getDefaultCameraProperties());
+}
+
+VRPNPlugin::~VRPNPlugin()
+{
+    _vrpnTracker.unregister_change_handler(&_camera, trackerCallback,
+                                           trackedSensorId);
+}
+
+void VRPNPlugin::preRender()
+{
+    _vrpnTracker.mainloop();
```

Comment:
this won't be called if brayns is idling, e.g. all accumulation frames are done. A wakeup with `api->triggerRender()` is required once the vrpn tracker has pending messages.

Reply:
`triggerRender` calls http://docs.libuv.org/en/v1.x/idle.html `start` which has no mentioning about thread-safety. It can be changed to an async handle, where the `send` is guaranteed to be thread-safe. Is the socket fd not exposed by vprn? Integrating it to libuv like DeflectPlugin does should work.

Reply:
No, I've been looking and I'm not even sure where the socket hides in the code. I think it's d_udpInboundSocket in vrpn_Endpoint_IP but it can't be accessed. :-(